### DISCUSSION TITLE
Anomaly detection

### DISCUSSION THREAD

**Discussion by codenicks** [TDAmeritrade/stumpy#622] [Q&A] [answered]
Hi, I am trying to use stumpy for univariate timeseries anomaly detection. I am still not sure if only discords represent anomalies or motifs too. I have a dataset like taxy example. Here is my discord discovery. I would like to know how to interpret this data and how I could get the index of the anomalies to plot them over the dataset with a different color. Any help is appreciated. Thankyou

![discord](https://user-images.githubusercontent.com/43618492/175093079-7e3d9e56-753c-4b47-9527-b7eeda7400ed.JPG)
!

**Comment by seanlaw** [comment, marked as answer]
@codenicks Welcome to the STUMPY community and thank you for your question. Are you able to provide the raw data and code for reproducing the plots above? That would really help

**Reply by codenicks** [reply]
yes, how can I send you the data ?

**Reply by seanlaw** [reply]
If you are able to save the data as a `.csv` file then you "should" be able to drag-and-drop it into the "Write a reply" text box below and it will automatically get uploaded and converted to a clickable link (see where it says "Attach files by dragging and dropping, selecting or pasting them"?).

**Reply by codenicks** [reply]
I didn't knew that, sorry. I have used the data_test_2022, which is an altered version of data_train_2021 with some random values.
The code is the same as in the examples with m=48.
Looking forward to hearing from you. Thankyou!

[data_train_2021.csv](https://github.com/TDAmeritrade/stumpy/files/8966619/data_train_2021.csv)
[data_test_2022.csv](https://github.com/TDAmeritrade/stumpy/files/8966620/data_test_2022.csv)

**Reply by seanlaw** [reply]
> I am still not sure if only discords represent anomalies or motifs too. 

Simply put, a motif and a discord are both obtained from a matrix profile. The only difference is that a motif, which has the smallest relative one-nearest-neighbor distance (in the matrix profile), is represented by the global minima while a discord, which has the largest relative one-nearest-neighbor distance , is represented by the global maxima. 

> I would like to know how to interpret this data and how I could get the index of the anomalies to plot them over the dataset with a different color. Any help is appreciated.

So, unlike finding a motif where we'd look for the global minima (i.e., global smallest value):

```
motif_idx = np.argsort(mp[:, 0])[0]
```

To find the global maxima (i.e., largest value), we'd do:

```
discord_idx = np.argsort(mp[:, 0])[-1]
```

Here, the `np.argsort(mp[:, 0])` simply sorts all of the matrix profile distances from smallest to largest and returns their corresponding index locations. Finally, the `[-1]` selects the index for the biggest matrix profile distance. Armed with this information, you can plot the discord (untested!!):

```
fig, axs = plt.subplots(2, sharex=True, gridspec_kw={'hspace': 0})
plt.suptitle('Discord (Anomaly) Discovery', fontsize='30')

axs[0].plot(data)  # Replace with your CSV raw data
rect = Rectangle((discord_idx, 0), m, 40, facecolor='lightgrey')
axs[0].add_patch(rect)
axs[1].set_xlabel('Time', fontsize ='20')
axs[1].set_ylabel('Matrix Profile', fontsize='20')
axs[1].axvline(x=discord_idx, linestyle="dashed")
axs[1].plot(mp[:, 0])
plt.show()
``` 

You can compare the similarities/differences with https://stumpy.readthedocs.io/en/latest/Tutorial_STUMPY_Basics.html?highlight=m%3D48#Find-a-Motif-Using-STUMP

**Reply by codenicks** [reply]
This is the code that I run before and I got the graph I posted in the question. Do you think the result looks ok? How should I interpret it ?

**Reply by seanlaw** [reply]
> This is the code that I run before and I got the graph I posted in the question. Do you think the result looks ok? 

Sure, the results look reasonable. I mean, politely, "it is what it is". 

> How should I interpret it?

By any chance, have you already watched [this video](https://stumpy.readthedocs.io/en/latest/motivation.html)? Starting around 16 minutes into it, it describes how to interpret the discords within a matrix profile. I think this video a good place to start. And then you can ask more specific questions.

**Reply by codenicks** [reply]
I have watched the video now, I also found about AB-Joins, maybe this is what I am looking for, I am not sure. Considering my data I think normal matrix profile without AB-Joins should also work. First thing that I would like to know is why that gray rectangle is not overlapping the graph. Also considering I found the top discord, how should I find the others ? Thankyou and I am sorry for so many questions.

**Reply by seanlaw** [reply]
> First thing that I would like to know is why that gray rectangle is not overlapping the graph. 

Ahh, it's because of `rect = Rectangle((discord_idx, 0), m, 40, facecolor='lightgrey')`. The rectangle is drawn by specifying the bottom-left-hand corner followed by the width (`m`) and height (`40`). So, the height needs to be changed accordingly. Perhaps, this is better:

```
rect = Rectangle((discord_idx, 0), m, max(data), facecolor='lightgrey')
``` 

This way, the rectangle is drawn and extends all the way to the top of the plot.

> Also considering I found the top discord, how should I find the others ? 

Once you find the first discord, you'll likely want to ignore any further discords in this area. So, what you can do is set the matrix profile values around each discord to, say, `np.NINF` (negative infinity):

```
mp = stumpy.stump(data, m)
discord_idx = np.argsort(mp[:, 0])[-1]
# Do something with this discord

# set last discord values to np.NINF
from stumpy import config
excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
zone_start = max(0, discord_idx - excl_zone)
zone_stop = min(mp[:, 0].shape[-1], discord_idx + excl_zone)
mp[zone_start : zone_stop + 1, 0] = np.NINF

# Find new discord just like before
discord_idx = np.argsort(mp[:, 0])[-1]
# Do something with this discord

# Repeat as needed
```

**Reply by codenicks** [reply]
For the rectangle I get this error if I do max(data)
![image](https://user-images.githubusercontent.com/43618492/175530721-b387c474-35a9-4ce5-8e78-9e0026a56c89.png)

Regarding the discords, is there a way to do a for loop until I get all discords ?

**Reply by seanlaw** [reply]
> For the rectangle I get this error if I do max(data)

I don't think your code is equivalent to `max(data)` (where `data` is your 1-D input to `stumpy.stump`). Based on line 4 of your code, the equivalent would be:

```
rect = Rectangle((discord_idx, 0), m, max(test_anomaly['T-HEX']), facecolor='lightgrey')
```

> Regarding the discords, is there a way to do a for loop until I get all discords ?

Yes, you'd do as I proposed above with the "repeat":

```
from stumpy import config

excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
mp = stumpy.stump(data, m)
discord_indices = []

# Get top 10 discords
for k in range(10):
    discord_idx = np.argsort(mp[:, 0])[-1]
    if mp[discord_idx, 0] == np.NINF:
        break

    # Do something with this discord
    discord_indices.append(discord_idx)

    # set discord values to np.NINF
    zone_start = max(0, discord_idx - excl_zone)
    zone_stop = min(mp[:, 0].shape[-1], discord_idx + excl_zone)
    mp[zone_start : zone_stop + 1, 0] = np.NINF
```

Something like that

**Reply by codenicks** [reply]
It is working, thankyou very much !
